src/transformStateWithClones.js

```diff
@@ -6,8 +6,29 @@
  *
  * @return {Object[]}
  */
+
 function transformStateWithClones(state, actions) {
-  // write code here
+  const stateHistory = [];
+  let currentState = { ...state };
+
+  actions.forEach((action) => {
+    if (action.type === 'clear') {
+      currentState = {};
+    } else if (action.type === 'addProperties') {
+      currentState = { ...currentState, ...action.extraData };
+    } else if (action.type === 'removeProperties') {
+      currentState = { ...currentState };
+
+      action.keysToRemove.forEach((key) => {
+        delete currentState[key];
+      });
+    }
```

Comment:
[CODE STYLE]: Use a switch statement if you have a limited amount of conditions. It is more readable and maintainable than multiple if/else statements when dealing with a type property. Additionally, ensure to include a default case for error handling.